### PR TITLE
HDDS-8378. [Snapshot] Integration test to simulate staggered snapshot creation and deletion

### PR DESCRIPTION
**Pull request by DaveTeng0**
## What changes were proposed in this pull request?
Integration test to simulate staggered snapshot creation and deletion

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-8378

## How was this patch tested?
Integration test.

### CODE REVIEW

**Comment by DaveTeng0** [comment]
cc. @smengcl @hemantk-12 @swamirishi @aswinshakil @prashantpogde

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmSnapshot.java [hunk 1]
```diff
@@ -1293,4 +1300,144 @@ public void testCompactionDagDisableForSnapshotMetadata() throws Exception {
...
+          monthSnapshotRetentionPeriodMonth, volumeA, bucketA, volAbucketA);
+      oldestMonthIndex = updatedMonthIndexArr[0];
+      latestMonthIndex = updatedMonthIndexArr[1];
+    }
+
+    checkDayWeekMonthSnapshotData(volAbucketA,
+        latestDayIndex,
+        latestWeekIndex,
+        latestMonthIndex,
+        KEY_TEST_DATA);
```
Comment (smengcl) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmSnapshot.java:1357]: We should verify data integrity each time a snapshot is taken, not just after the loop.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmSnapshot.java [hunk 2]
```diff
@@ -1293,4 +1300,144 @@ public void testCompactionDagDisableForSnapshotMetadata() throws Exception {
...
+            OmSnapshotManager.getSnapshotPrefix(snapshotPrefix + (index - 1)) +
+                KEY_PREFIX);
+    boolean findKey = false;
+    while (iterator.hasNext()) {
+      OzoneKey ozoneKey = iterator.next();
+      if (ozoneKey.getName().contains(keyName)) {
+        findKey = true;
+        break;
+      }
+    }
```
Comment (smengcl) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmSnapshot.java:1429]: In a follow-up patch, we should check the actual content written to the key as well.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmSnapshot.java [hunk 3]
```diff
@@ -1293,4 +1300,144 @@ public void testCompactionDagDisableForSnapshotMetadata() throws Exception {
...
+      int oldestIndex, int latestIndex, int snapshotRetentionPeriod,
+      String volumeNameStr, String bucketNameStr, OzoneBucket ozoneBucketClient)
+      throws Exception {
+    if (latestIndex - oldestIndex >= snapshotRetentionPeriod) {
+      store.deleteSnapshot(volumeNameStr, bucketNameStr,
+          snapshotPrefix + oldestIndex);
+      oldestIndex++;
+    }
+    createFileKeyWithData(ozoneBucketClient, KEY_PREFIX + latestIndex,
+        KEY_TEST_DATA);
```
Comment (smengcl) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmSnapshot.java:1370]: TODO: Write different test data for each key

**Review by prashantpogde — APPROVED** [review]
We need to have another incremental patch on top of this as suggested in the comment.
- verify data integrity each time a snapshot is taken
- different test data for each key
- content verification

I am approving this PR as base PR to prevent PR from becoming too big with further enhancements.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmSnapshot.java (hunk 3 above)
Comment (DaveTeng0) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmSnapshot.java]: sure, update.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmSnapshot.java (hunk 2 above)
Comment (DaveTeng0) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmSnapshot.java]: sure, updated

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmSnapshot.java (hunk 1 above)
Comment (DaveTeng0) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmSnapshot.java]: sure, updated. Checked key's content after each snapshot deletion.

**Comment by prashantpogde** [comment]
Thank you for this PR @DaveTeng0

**Pull request merged**